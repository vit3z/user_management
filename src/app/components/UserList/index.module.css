.userListDisplayLoadingContainer {
  margin: 24px auto;
  display: flex;
  justify-content: center;
}

.searchContainer {
  margin-top: 8px;
}

.userDisplaySearchBar {
  height: 32px;
  border-radius: 8px;
  border: 1px solid aliceblue;
  width: 250px;
  padding: 0px 4px;
}

.userDisplaySearchBar:focus { 
  outline: 1px solid#17D3FF !important;
}

.userListDisplayContainer {
  width: 100%;
  margin-top: 8px;
  background-color: aliceblue;
  border-radius: 8px;
}

.userDisplayBasicInfo {
  height: 32px;
  display: flex;
  justify-content: space-between;
  flex-direction: row;
  align-items: center;
  margin: 8px 0px;
  cursor: pointer;
  padding: 0px 4px;
}

.userDisplayBasicInfo:hover {
  background-color: rgb(229, 237, 245);
  -webkit-transition: background-color 100ms ease-in;
  -ms-transition: background-color 100ms ease-in;
  transition: background-color 100ms ease-in;
}

.userDisplayBasicInfoTitle {
  border-top-left-radius: 8px;
  border-top-right-radius: 8px;
  background-color: #FFA817;
  height: 32px;
  display: flex;
  justify-content: space-between;
  flex-direction: row;
  font-weight: bold;
  padding: 0px 4px;
  align-items: center;
}

.userListUserAvatar {
  border-radius: 50%;
  margin-left: 7px;
  margin-right: -7px;
}

.userDisplayItemBlank {
  width: 24px;
}

.userDisplayItemTitle {
  cursor: pointer;
  width: 30%;
}
.userDisplayItem {
  width: 30%;
}

.userListErrorMessage {
  color: #FF5C17;
  margin: 8px auto;
}

.noUsersFound {
  margin: 8px;
}

@media (max-width: 768px) {
  .userDisplaySearchBar {
    width: 100%;
  }

  .userListDisplayContainer {
    width: 100%;
    margin-top: 16px;
    background-color: rgba(0, 0, 0, 0);
    border-radius: 8px;
  }

  .userListUserAvatar {
    margin-top: 4px;
    margin-left: 0px;
    margin-right: 0px;
  }

  .userDisplayBasicInfo {
    height: 120px;
    flex-direction: column;
    background-color: aliceblue;
    justify-content: space-around;
    border-radius: 8px;
    border-bottom: 1px solid darkgray;
    border-right: 1px solid darkgray;
    border-top: 1px solid rgb(228, 228, 228);
    border-left: 1px solid rgb(228, 228, 228);
    padding-bottom: 4px;
  }

  .userDisplayBasicInfoTitle {
    display: none;
  }

  .userDisplayItem {
    display: flex;
    justify-content: center;
    width: 100%;
    overflow: hidden;
    text-overflow: ellipsis;
  }  
}