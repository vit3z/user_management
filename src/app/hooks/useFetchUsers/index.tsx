import { useState, useEffect } from "react";

const useFetchUsers = () => {
  const [users, setUsers] = useState<any>(null);
  const [error, setError] = useState<string | null>(null);
  const [loading, setLoading] = useState<boolean>(true);

  useEffect(() => {
    const fetchUsers = async () => {
      try {
        const response = await fetch(
          `https://jsonplaceholder.typicode.com/users`,
          {
            method: "GET",
            headers: {
              "Content-Type": "application/json",
            },
          }
        );

        if (!response.ok) {
          throw new Error("User fetch failed");
        }

        const userList = await response.json();
        setUsers(userList);
      } catch (error) {
        setError(`${error}`);
      } finally {
        setLoading(true);
      }
    };
    fetchUsers();
  }, []);

  return { users, error, loading };
};

export default useFetchUsers;
