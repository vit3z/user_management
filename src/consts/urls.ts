export const userFetchUrl = `https://jsonplaceholder.typicode.com/users`;