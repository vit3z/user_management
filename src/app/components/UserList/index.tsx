"use client";

import React, { useEffect, useState } from "react";
import styles from "./index.module.css";
import useFetchUsers from "@/app/hooks/useFetchUsers";
import { UserInformationProps } from "@/consts/interfaces";
import LoadingSpinner from "../LoadingSpinner";
import UserDetails from "../UserDetails";
import Image from "next/image";
import ProfileImage from "../../../../public/Profile_avatar_placeholder_large.png";
import { columnNameGenerator } from "@/app/utils/ColumnNameGenerator";

const UserList: React.FC = () => {
  const [isModalOpen, setModalOpen] = useState<boolean>(false);
  const [userDetails, setUserDetails] = useState<UserInformationProps | null>(
    null
  );
  const [userData, setUserData] = useState<UserInformationProps[] | null>(null);
  const [currentSort, setCurrentSort] = useState<string>("");

  const { users, error, loading } = useFetchUsers();

  const handleOpenModal = (user: UserInformationProps) => {
    setUserDetails(user);
    setModalOpen(true);
  };

  const handleCloseModal = () => {
    setUserDetails(null);
    setModalOpen(false);
  };

  useEffect(() => {
    setUserData(users);
  }, [users]);

  const handleSortBy = (sortParam: string) => {
    const clonedUserList = JSON.parse(JSON.stringify(userData));
    switch (sortParam) {
      case "name":
        clonedUserList.sort(
          (a: UserInformationProps, b: UserInformationProps) => {
            let nameA = a.name.toLowerCase();
            let nameB = b.name.toLowerCase();

            if (nameA > nameB) return 1;
            if (nameA < nameB) return -1;
            return 0;
          }
        );
        if(currentSort !== "nameAsc" && currentSort !== "nameDesc") {
          setCurrentSort("nameAsc");
        }
        if(currentSort === "nameAsc") {
          setCurrentSort("nameDesc");
          clonedUserList.reverse();
        }
        if(currentSort === "nameDesc") {
          setCurrentSort("nameAsc");
        }
        break;
      case "email":
        clonedUserList.sort(
          (a: UserInformationProps, b: UserInformationProps) => {
            let emailA = a.email.toLowerCase();
            let emailB = b.email.toLowerCase();

            if (emailA > emailB) return 1;
            if (emailA < emailB) return -1;
            return 0;
          }
        );
        if(currentSort !== "emailAsc" && currentSort !== "emailDesc") {
          setCurrentSort("emailAsc");
        }
        if(currentSort === "emailAsc") {
          setCurrentSort("emailDesc");
          clonedUserList.reverse();
        }
        if(currentSort === "emailDesc") {
          setCurrentSort("emailAsc");
        }
        break;
      case "phoneNum":
        clonedUserList.sort(
          (a: UserInformationProps, b: UserInformationProps) => {
            let phoneA = a.phone.toLowerCase();
            let phoneB = b.phone.toLowerCase();

            if (phoneA > phoneB) return 1;
            if (phoneA < phoneB) return -1;
            return 0;
          }
        );
        if(currentSort !== "phoneAsc" && currentSort !== "phoneDesc") {
          setCurrentSort("phoneAsc");
        }
        if(currentSort === "phoneAsc") {
          setCurrentSort("phoneDesc");
          clonedUserList.reverse();
        }
        if(currentSort === "phoneDesc") {
          setCurrentSort("phoneAsc");
        }
        break;
      default:
        break;
    }
    setUserData(clonedUserList);
  };

  const handleSearchForUser = (e: React.ChangeEvent<HTMLInputElement>) => {
    if(e.target.value !== "") {
      const newResults = users?.filter((el: UserInformationProps) => {
        return el.name.toLowerCase().includes(e.target.value.toLowerCase());
      })
      if(newResults && newResults?.length > 0) {
        setUserData(newResults);
      } else {
        setUserData([]);
      }
      setCurrentSort("");
    } else {
      setUserData(users);
    }
  }

  return (
    <>
      <div className={styles.searchContainer}>
        <span className={styles.searchContainerIcon}>&#x1F50E;&#xFE0E;</span>
        <input className={styles.userDisplaySearchBar} placeholder="Search User By Name" onChange={(e) => handleSearchForUser(e)}/>
      </div>
      <div className={styles.userListDisplayContainer}>
        {!loading && !error && (
          <ul>
            <div className={styles.userDisplayBasicInfoTitle}>
              <p className={styles.userDisplayItemBlank}>&nbsp;</p>
              <p
                className={styles.userDisplayItemTitle}
                onClick={() => handleSortBy("name")}
              >
                {columnNameGenerator("Name", currentSort)}
              </p>
              <p
                className={styles.userDisplayItemTitle}
                onClick={() => handleSortBy("email")}
              >
                {columnNameGenerator("Email", currentSort)}
              </p>
              <p
                className={styles.userDisplayItemTitle}
                onClick={() => handleSortBy("phoneNum")}
              >
                {columnNameGenerator("Phone Number", currentSort)}
              </p>
            </div>
            {userData?.map((user: UserInformationProps, index: number) => {
              return (
                <div
                  className={index%2===0 ? styles.userDisplayBasicInfoEven : styles.userDisplayBasicInfoOdd}
                  key={`user-management-list-${index}-${user.id}`}
                  onClick={() => handleOpenModal(user)}
                >
                  <Image
                    className={styles.userListUserAvatar}
                    src={ProfileImage}
                    width={24}
                    height={24}
                    alt="User Avatar"
                  />
                  <p className={styles.userDisplayItem}>{user.name}</p>
                  <p className={styles.userDisplayItem}>{user.email}</p>
                  <p className={styles.userDisplayItem}>{user.phone}</p>
                </div>
              );
            })}
            {(userData && userData?.length < 1) && <p className={styles.noUsersFound}>No Users Found</p>}
          </ul>
        )}
        {loading && (
          <div className={styles.userListDisplayLoadingContainer}>
            <LoadingSpinner />
          </div>
        )}
        {error && <p className={styles.userListErrorMessage}>{error}</p>}

        {userDetails && (
          <UserDetails
            isOpen={isModalOpen}
            onClose={handleCloseModal}
            data={userDetails}
          />
        )}
      </div>
    </>
  );
};

export default UserList;
