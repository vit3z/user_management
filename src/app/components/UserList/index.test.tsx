import "@testing-library/jest-dom";
import {
  render,
  fireEvent,
  cleanup,
  waitFor,
  screen,
} from "@testing-library/react";
import UserList from ".";
import useFetchUsers from "../../hooks/useFetchUsers";
import { UserInformationProps } from "@/consts/interfaces";
jest.mock("../../hooks/useFetchUsers");

describe("UserList", () => {
  const mockUsers = [
    {
      id: 1,
      name: "Leanne Graham",
      username: "Bret",
      email: "Sincere@april.biz",
      address: {
        street: "Kulas Light",
        suite: "Apt. 556",
        city: "Gwenborough",
        zipcode: "92998-3874",
        geo: {
          lat: "-37.3159",
          lng: "81.1496",
        },
      },
      phone: "1-770-736-8031 x56442",
      website: "hildegard.org",
      company: {
        name: "Romaguera-Crona",
        catchPhrase: "Multi-layered client-server neural-net",
        bs: "harness real-time e-markets",
      },
    },
    {
      id: 2,
      name: "Ervin Howell",
      username: "Antonette",
      email: "Shanna@melissa.tv",
      address: {
        street: "Victor Plains",
        suite: "Suite 879",
        city: "Wisokyburgh",
        zipcode: "90566-7771",
        geo: {
          lat: "-43.9509",
          lng: "-34.4618",
        },
      },
      phone: "010-692-6593 x09125",
      website: "anastasia.net",
      company: {
        name: "Deckow-Crist",
        catchPhrase: "Proactive didactic contingency",
        bs: "synergize scalable supply-chains",
      },
    },
  ] as UserInformationProps[];

  const mockUseFetchUsers = jest.mocked(useFetchUsers);

  beforeEach(() => {
    mockUseFetchUsers.mockReturnValue({
      users: mockUsers,
      error: null,
      loading: false,
    });
  });
  it("should render the component", () => {
    const { queryByPlaceholderText, container } = render(<UserList />);

    const searchBar = queryByPlaceholderText("Search User By Name");
    expect(searchBar).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should display user names", async () => {
    const { getByText } = render(<UserList />);

    await waitFor(() => {
      const user1 = getByText("Leanne Graham");
      const user2 = getByText("Ervin Howell");

      expect(user1).toBeInTheDocument();
      expect(user2).toBeInTheDocument();
    });
  });

  it("should search users by name", async () => {
    const { queryByText } = render(<UserList />);

    fireEvent.change(screen.getByPlaceholderText("Search User By Name"), {
      target: { value: "Leanne" },
    });

    await waitFor(() => {
      const user1 = queryByText("Leanne Graham");
      const user2 = queryByText("Ervin Howell");

      expect(user1).toBeInTheDocument();
      expect(user2).not.toBeInTheDocument();
    });
  });

  it("should return an empty result when looking for a user that is not in the list", async () => {
    const { queryByText } = render(<UserList />);

    fireEvent.change(screen.getByPlaceholderText("Search User By Name"), {
      target: { value: "Test" },
    });

    await waitFor(() => {
      const user1 = queryByText("Leanne Graham");
      const user2 = queryByText("Ervin Howell");
      const noUser = queryByText("No Users Found");

      expect(user1).not.toBeInTheDocument();
      expect(user2).not.toBeInTheDocument();
      expect(noUser).toBeInTheDocument();
    });
  });

  it("should return the list to it's initial state when there is no text in the search bar", async () => {
    const { queryByText } = render(<UserList />);

    fireEvent.change(screen.getByPlaceholderText("Search User By Name"), {
      target: { value: "Test" },
    });
    fireEvent.change(screen.getByPlaceholderText("Search User By Name"), {
      target: { value: "" },
    });

    await waitFor(() => {
      const user1 = queryByText("Leanne Graham");
      const user2 = queryByText("Ervin Howell");

      expect(user1).toBeInTheDocument();
      expect(user2).toBeInTheDocument();
    });
  });

  it("should sort users by name in asc order when the appropriate element is clicked", async () => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Name"));

    await waitFor(() => {
      const sortedUsers = screen.getAllByText(/Ervin Howell|Leanne Graham/);
      expect(sortedUsers[0]).toHaveTextContent("Ervin Howell");
      expect(sortedUsers[1]).toHaveTextContent("Leanne Graham");
    });
  });

  it("should sort users by name in desc order when the appropriate element is clicked twice", async () => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Name"));
    fireEvent.click(screen.getByText("Name ▲"));

    await waitFor(() => {
      const sortedUsers = screen.getAllByText(/Ervin Howell|Leanne Graham/);
      expect(sortedUsers[0]).toHaveTextContent("Leanne Graham");
      expect(sortedUsers[1]).toHaveTextContent("Ervin Howell");
    });
  });

  it("should sort users by name in asc order when the appropriate element is clicked thrice", async () => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Name"));
    fireEvent.click(screen.getByText("Name ▲"));
    fireEvent.click(screen.getByText("Name ▼"));

    await waitFor(() => {
      const sortedUsers = screen.getAllByText(/Ervin Howell|Leanne Graham/);
      expect(sortedUsers[0]).toHaveTextContent("Ervin Howell");
      expect(sortedUsers[1]).toHaveTextContent("Leanne Graham");
    });
  });

  it("should sort users by email in asc order when the appropriate element is clicked", async () => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Email"));

    await waitFor(() => {
      const sortedUsers = screen.getAllByText(
        /Shanna@melissa.tv|Sincere@april.biz/
      );
      expect(sortedUsers[0]).toHaveTextContent("Shanna@melissa.tv");
      expect(sortedUsers[1]).toHaveTextContent("Sincere@april.biz");
    });
  });

  it("should sort users by email in desc order when the appropriate element is clicked twice", async () => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Email"));
    fireEvent.click(screen.getByText("Email ▲"));

    await waitFor(() => {
      const sortedUsers = screen.getAllByText(
        /Shanna@melissa.tv|Sincere@april.biz/
      );
      expect(sortedUsers[0]).toHaveTextContent("Sincere@april.biz");
      expect(sortedUsers[1]).toHaveTextContent("Shanna@melissa.tv");
    });
  });

  it("should sort users by email in asc order when the appropriate element is clicked thrice", async () => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Email"));
    fireEvent.click(screen.getByText("Email ▲"));
    fireEvent.click(screen.getByText("Email ▼"));

    await waitFor(() => {
      const sortedUsers = screen.getAllByText(
        /Shanna@melissa.tv|Sincere@april.biz/
      );
      expect(sortedUsers[0]).toHaveTextContent("Shanna@melissa.tv");
      expect(sortedUsers[1]).toHaveTextContent("Sincere@april.biz");
    });
  });

  it("should sort users by phone number in asc order when the appropriate element is clicked", async () => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Phone Number"));

    await waitFor(() => {
      const sortedUsers = screen.getAllByText(
        /010-692-6593 x09125|1-770-736-8031 x56442/
      );
      expect(sortedUsers[0]).toHaveTextContent("010-692-6593 x09125");
      expect(sortedUsers[1]).toHaveTextContent("1-770-736-8031 x56442");
    });
  });

  it("should sort users by phone number in desc order when the appropriate element is clicked twice", async () => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Phone Number"));
    fireEvent.click(screen.getByText("Phone Number ▲"));

    await waitFor(() => {
      const sortedUsers = screen.getAllByText(
        /010-692-6593 x09125|1-770-736-8031 x56442/
      );
      expect(sortedUsers[0]).toHaveTextContent("1-770-736-8031 x56442");
      expect(sortedUsers[1]).toHaveTextContent("010-692-6593 x09125");
    });
  });

  it("should sort users by phone number in asc order when the appropriate element is clicked thrice", async () => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Phone Number"));
    fireEvent.click(screen.getByText("Phone Number ▲"));
    fireEvent.click(screen.getByText("Phone Number ▼"));

    await waitFor(() => {
      const sortedUsers = screen.getAllByText(
        /010-692-6593 x09125|1-770-736-8031 x56442/
      );
      expect(sortedUsers[0]).toHaveTextContent("010-692-6593 x09125");
      expect(sortedUsers[1]).toHaveTextContent("1-770-736-8031 x56442");
    });
  });

  it("should open the details modal when the user clicks on a data point", async () => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Leanne Graham"));

    await waitFor(() => {
      const modalUserName = screen.queryByText("Bret");

      expect(modalUserName).toBeInTheDocument();
    });
  });

  it("should handle the closing of the modal", async() => {
    render(<UserList />);

    fireEvent.click(screen.getByText("Leanne Graham"));

    await waitFor(() => {
      const modalUserName = screen.queryByText("Bret");
      expect(modalUserName).toBeInTheDocument();
      fireEvent.click(screen.getByText("Close"));
      expect(modalUserName).not.toBeInTheDocument();
    });
  });
});
