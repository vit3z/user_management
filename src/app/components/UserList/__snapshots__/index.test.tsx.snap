// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList should render the component 1`] = `
<div>
  <div
    class="searchContainer"
  >
    <span
      class="searchContainerIcon"
    >
      ðŸ”Žï¸Ž
    </span>
    <input
      class="userDisplaySearchBar"
      placeholder="Search User By Name"
    />
  </div>
  <div
    class="userListDisplayContainer"
  >
    <ul>
      <div
        class="userDisplayBasicInfoTitle"
      >
        <p
          class="userDisplayItemBlank"
        >
          Â 
        </p>
        <p
          class="userDisplayItemTitle"
        >
          <span>
            NAME â‡…
          </span>
        </p>
        <p
          class="userDisplayItemTitle"
        >
          <span>
            USERNAME â‡…
          </span>
        </p>
        <p
          class="userDisplayItemTitle"
        >
          <span>
            EMAIL â‡…
          </span>
        </p>
        <p
          class="userDisplayItemTitle"
        >
          <span>
            PHONE NUMBER â‡…
          </span>
        </p>
      </div>
      <div
        class="userDisplayBasicInfoEven"
      >
        <img
          alt="User Avatar"
          class="userListUserAvatar"
          data-nimg="1"
          decoding="async"
          height="24"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
          style="color: transparent;"
          width="24"
        />
        <p
          class="userDisplayItem"
        >
          Leanne Graham
        </p>
        <p
          class="userDisplayItem"
        >
          Bret
        </p>
        <p
          class="userDisplayItem"
        >
          Sincere@april.biz
        </p>
        <p
          class="userDisplayItem"
        >
          1-770-736-8031 x56442
        </p>
      </div>
      <div
        class="userDisplayBasicInfoOdd"
      >
        <img
          alt="User Avatar"
          class="userListUserAvatar"
          data-nimg="1"
          decoding="async"
          height="24"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
          style="color: transparent;"
          width="24"
        />
        <p
          class="userDisplayItem"
        >
          Ervin Howell
        </p>
        <p
          class="userDisplayItem"
        >
          Antonette
        </p>
        <p
          class="userDisplayItem"
        >
          Shanna@melissa.tv
        </p>
        <p
          class="userDisplayItem"
        >
          010-692-6593 x09125
        </p>
      </div>
    </ul>
  </div>
</div>
`;
